### PR TITLE
[REVIEW] Add python methods to api.rst, fix formatting

### PR DESCRIPTION
**Pull request by trxcllnt**
This PR fixes a few formatting issues in our Python docs and adds all the new methods to the Sphinx `api.rst` manifest.

If you're using `rapids-compose`, here's an easy way to build the python docs:
```shell
make html -C $CUSPATIAL_HOME/docs
```

Or if you're doing a lot of docs editing and want to auto-rebuild the docs on save, you can do this:
```shell
# Install `entr` to watch files (https://systutorials.com/docs/linux/man/1-entr/)
echo "rapids" | sudo -S apt update && echo "rapids" | sudo -S apt install entr

# Launch `entr` to re-run `make html` any time the source or config files change
find $CUSPATIAL_HOME/python $CUSPATIAL_HOME/docs/source \
  | entr sh -c "make html -C $CUSPATIAL_HOME/docs"
```

<details><summary>Expand this to see a screenshot of the rendered docs</summary>

</details>

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

File: python/cuspatial/cuspatial/core/interpolate.py [hunk 1]
```diff
@@ -46,16 +46,16 @@ class CubicSpline:
...
         y = cudf.Series(np.random.random(100*1000)).astype('float32')
         prefix_sum = cudf.Series(cp.arange(1000)*100).astype('int32')
-        new_samples = cudf.Series(np.repeat(np.linspace(0, 100, 1000), 1000)
-            .astype('float32'))
+        new_samples = cudf.Series(
+            np.repeat(np.linspace(0, 100, 1000), 1000)
+        ).astype('float32')
 
         curve = cuspatial.CubicSpline(t, y, prefixes=prefix_sum)
         new_points = curve(new_samples, prefix_sum*10)
```
Comment (trxcllnt) [inline python/cuspatial/cuspatial/core/interpolate.py:61]: I tried running this example in a notebook, but I'm getting CUDA illegal address errors. Does this need updating for any recent API or implementation changes?

File: python/cuspatial/cuspatial/core/gis.py [hunk 2]
```diff
@@ -84,19 +92,23 @@ def haversine_distance(p1_lon, p1_lat, p2_lon, p2_lat):
...
+        longitude of second set of coords
+    p2_lat
+        latitude of second set of coords
 
-    returns
-    Series: distance between all pairs of lat/lon coords
+    Returns
+    -------
+    result : cudf.Series
+        The distance between all pairs of lat/lon coordinates
```
Comment (harrism) [inline python/cuspatial/cuspatial/core/gis.py:109]: Order seems to switch here?

```suggestion
        The distance between all pairs of lon/lat coordinates
```

**Comment by trxcllnt** [comment]
@thomcom @cwharris do you guys have feedback for this PR?

File: python/cuspatial/cuspatial/core/gis.py [hunk 3]
```diff
@@ -21,53 +21,61 @@ def directed_hausdorff_distance(xs, ys, points_per_space):
...
+        column of y-coordinates
+    points_per_space
+        number of points in each space
 
-    Example
+    Returns
     -------
+    result : cudf.DataFrame
+        The pairwise directed distance matrix with one row and one
+        column per input space; the value at row i, column j represents the
```
Comment (cwharris) [inline python/cuspatial/cuspatial/core/gis.py:37]: Did we verify this is correct (aka not vice versa)? I never did. I wrote it so the existing tests would pass, and I believe the existing docs state it this way, but I never confirmed the existing docs were correct. :/

File: python/cuspatial/cuspatial/core/gis.py [hunk 4]
```diff
@@ -21,53 +21,61 @@ def directed_hausdorff_distance(xs, ys, points_per_space):
...
-    [Wikipedia](https://en.wikipedia.org/wiki/Hausdorff_distance)
+    `Wikipedia <https://en.wikipedia.org/wiki/Hausdorff_distance>`_
+
+    Consider a pair of lines on a grid::
+
+             :
+             x
+        -----xyy---
+             :
+             :
```
Comment (cwharris) [inline python/cuspatial/cuspatial/core/gis.py:54]: What does this have to do with Hausdorff?

File: python/cuspatial/cuspatial/core/gis.py [hunk 5]
```diff
@@ -107,21 +119,28 @@ def haversine_distance(p1_lon, p1_lat, p2_lon, p2_lat):
...
-    origin.
-
-    params
-    origin_lon: float64 - longitude camera
-    origin_lat: float64 - latitude camera
-    input_lon: Series of longitude coords to convert to x
-    input_lat: Series of latitude coords to convert to y
+    """
+    Convert lon/lat coordinates to kilometer x,y coordinates with respect to
+    an origin camera position.
```
Comment (cwharris) [inline python/cuspatial/cuspatial/core/gis.py:124]: Convert lon/lat to x,y coordinates relative to an origin lon/lat. Results are scaled relative to the size of Earth in kilometers.

File: python/cuspatial/cuspatial/core/gis.py [hunk 6]
```diff
@@ -107,21 +119,28 @@ def haversine_distance(p1_lon, p1_lat, p2_lon, p2_lat):
...
 
     Parameters
     ----------
-    {params}
+    origin_lon : ``number``
+        camera longitude
+    origin_lat : ``number``
+        camera latitude
```
Comment (cwharris) [inline python/cuspatial/cuspatial/core/gis.py:131]:
```suggestion
    origin_lon : ``number``
        longitude offset  (this is subtracted from each input before converting to x,y)
    origin_lat : ``number``
        latitude offset (this is subtracted from each input before converting to x,y)
```

File: python/cuspatial/cuspatial/core/gis.py [hunk 7]
```diff
@@ -107,21 +119,28 @@ def haversine_distance(p1_lon, p1_lat, p2_lon, p2_lat):
...
 
-    returns
-    DataFrame: 'x', 'y' columns for new km positions of coords
+    Returns
+    -------
+    result : cudf.DataFrame
+        x : cudf.Series
+            x-coordinate of the new positions in kilometers
+        y : cudf.Series
+            y-coordinate of the new positions in kilometers
```
Comment (cwharris) [inline python/cuspatial/cuspatial/core/gis.py:143]:
```suggestion
    result : cudf.DataFrame
        x : cudf.Series
            x-coordinates of the input relative to the size of Earth in kilometers.
        y : cudf.Series
            y-coordinate of the input relative to the size of Earth in kilometers
```

File: python/cuspatial/cuspatial/core/gis.py [hunk 8]
```diff
@@ -142,51 +161,55 @@ def point_in_polygon(
...
+
+    Test whether 3 points fall within either of two polygons
+
+    >>> result = cuspatial.point_in_polygon(
+        [0, -8, 6.0],                             # test_points_x
+        [0, -8, 6.0],                             # test_points_y
+        cudf.Series([0, 1], index=['nyc', 'dc']), # poly_offsets
+        [0, 3],                                   # ring_offsets
+        [-10, 5, 5, -10, 0, 10, 10, 0],           # poly_points_x
+        [-10, -10, 5, 5, 0, 0, 10, 10],           # poly_points_y
```
Comment (cwharris) [inline python/cuspatial/cuspatial/core/gis.py:190]: I believe the first point in each ring must be repeated at the end of the ring (completing the loop). This is contrary to the c++ documentation. My initial tests seemed to indicate otherwise, but I believe there are situations when the first points must be repeated.  Specifically in the case when the "missing edge" should intersects an x-axis determined by the y-coordinate of the input point. This still needs to be verified. For example:

```
// 3 -> 0 may need to be explicitly indicated in the ring by repeating the first point at the end of the ring.
// the test for point A will be correct, but tests for point B and C will be incorrect.

ring points: [0, 1, 2, 3]

   0-----1
         |
A     B  |  C
         |
   3-----2
```

This didn't come up in initial testing because the ring's last edge was aligned with the x-axis:
```
ring points: [3, 0, 1, 2]:

   0-----1
   |     |
A  |  B  |  C
   |     |
   3     2
```

File: python/cuspatial/cuspatial/core/gis.py [hunk 9]
```diff
@@ -142,51 +161,55 @@ def point_in_polygon(
...
-        # Point 0: (0, 0) falls in both polygons
-        # Point 1: (-8, -8) falls in the first polygon
-        # Point 2: (6.0, 6.0) falls in the second polygon
+
+    Test whether 3 points fall within either of two polygons
+
+    >>> result = cuspatial.point_in_polygon(
+        [0, -8, 6.0],                             # test_points_x
+        [0, -8, 6.0],                             # test_points_y
+        cudf.Series([0, 1], index=['nyc', 'dc']), # poly_offsets
```
Comment (cwharris) [inline python/cuspatial/cuspatial/core/gis.py:187]: Maybe use hudson river in place of dc, since it partially overlaps with nyc (may be less confusing).
```suggestion
        [0, -8, 6.0],                             # test_points_y
        cudf.Series([0, 1], index=['nyc', 'hudson river']), # poly_offsets
```

File: python/cuspatial/cuspatial/core/gis.py [hunk 10]
```diff
@@ -142,51 +161,55 @@ def point_in_polygon(
...
+        [0, 3],                                   # ring_offsets
+        [-10, 5, 5, -10, 0, 10, 10, 0],           # poly_points_x
+        [-10, -10, 5, 5, 0, 0, 10, 10],           # poly_points_y
+    )
+    # The result of point_in_polygon is a DataFrame of Boolean
+    # values indicating whether each point (rows) falls within
+    # each polygon (columns).
+    >>> print(result)
+                nyc            dc
```
Comment (cwharris) [inline python/cuspatial/cuspatial/core/gis.py:196]:
```suggestion
                nyc            hudson river
```

File: python/cuspatial/cuspatial/core/interpolate.py (hunk 1 above)
Comment (cwharris) [reply python/cuspatial/cuspatial/core/interpolate.py:61]: @thomcom

File: python/cuspatial/cuspatial/core/gis.py (hunk 4 above)
Comment (trxcllnt) [reply python/cuspatial/cuspatial/core/gis.py:54]: I don't know, I assumed you wrote this diagram? I just corrected the formatting.

File: python/cuspatial/cuspatial/core/gis.py (hunk 5 above)
Comment (trxcllnt) [reply python/cuspatial/cuspatial/core/gis.py:124]: please make a github suggestion as to what to change?

File: python/cuspatial/cuspatial/io/shapefile.py [hunk 11]
```diff
@@ -8,7 +8,28 @@
...
+        ESRI Shapefile file path (usually ends in ``.shp``)
+
+    Returns
+    -------
+    result  : tuple (cudf.Series, cudf.Series, cudf.DataFrame)
+    poly_offsets   : cudf.Series(dtype=np.int32)
+        Offsets of the first ring in each polygon
+    ring_offsets   : cudf.Series(dtype=np.int32)
+        Offsets of the first point in each ring
+    points  : cudf.DataFrame
```
Comment (cwharris) [inline python/cuspatial/cuspatial/io/shapefile.py:26]: It's strange that we don't return a single DataFrame, or four Series... But this is a documentation PR. :)

File: python/cuspatial/cuspatial/core/trajectory.py [hunk 12]
```diff
@@ -92,35 +131,49 @@ def trajectory_bounding_boxes(num_trajectories, object_ids, xs, ys):
...
     ----------
-    {params}
+    num_trajectories
+        number of trajectories (unique object ids)
+    object_ids
+        column of object (e.g., vehicle) ids
+    xs
+        column of x-coordinates (in kilometers)
+    ys
+        column of y-coordinates (in kilometers)
```
[the remark this answers is not in the document]
Comment (trxcllnt) [reply python/cuspatial/cuspatial/core/trajectory.py:146]: The inputs must be in kilometers if meters/second is the desired output. Though if the inputs were in miles, the output speeds would be milli-miles/second since we're just dividing the input units 1000.

File: python/cuspatial/cuspatial/io/shapefile.py (hunk 11 above)
Comment (trxcllnt) [reply python/cuspatial/cuspatial/io/shapefile.py:26]: I believe all these could be different lengths?

File: python/cuspatial/cuspatial/core/gis.py (hunk 4 above)
Comment (cwharris) [reply python/cuspatial/cuspatial/core/gis.py:54]: I don’t believe I wrote these docs. Doesn't seem to be relevant to the rest of the example.

File: python/cuspatial/cuspatial/core/trajectory.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (cwharris) [reply python/cuspatial/cuspatial/core/trajectory.py:146]: Ah, I see. Perhaps We should leave the results in kilometers/s, or add an optional multiplier? Not relevant this PR, though.

File: python/cuspatial/cuspatial/io/shapefile.py (hunk 11 above)
Comment (cwharris) [reply python/cuspatial/cuspatial/io/shapefile.py:26]: Right. That makes sense. 🤦
